src/openms_gui/source/VISUAL/TOPPViewIdentificationViewBehavior.cpp

```diff
@@ -894,12 +833,143 @@ namespace OpenMS
     widget_1D->canvas()->setTextBox(QString());
   }
 
+  void TOPPViewIdentificationViewBehavior::addAnnotationsSpectrumLayer_(const PeptideHit& hit, bool align)
+  {
+    const vector<PeptideHit::PeakAnnotation>& annotations =
+      hit.getPeakAnnotations();
+    String seq = hit.getSequence().toString();
+    if (seq.empty()) seq = hit.getMetaValue("label");
+
+    SpectrumCanvas* current_canvas = tv_->getActive1DWidget()->canvas();
+    LayerData& current_layer = current_canvas->getCurrentLayer();
+    Size current_spectrum_layer_index = current_canvas->activeLayerIndex();
+    Size current_spectrum_index = current_layer.getCurrentSpectrumIndex();
+
+    const MSSpectrum& current_spectrum = current_layer.getCurrentSpectrum();
+    if (align)
+    {
+      if (current_spectrum.empty())
+      {
+        LOG_WARN << "Spectrum is empty! Nothing to annotate!" << std::endl;
+      }
+      else if (!current_spectrum.isSorted())
+      {
+        QMessageBox::warning(tv_, "Error", "The spectrum is not sorted! Aborting!"); // @TODO: improve error message
+        return;
+      }
+    }
+
+    MSSpectrum ann_spectrum;
+    vector<String> labels;
+    for (const auto& ann : annotations)
+    {
+      Peak1D peak(ann.mz, ann.intensity);
+      if (align) // align to the measured spectrum
+      {
+        // @TODO: avoid magic constant (m/z tolerance)
+        Int peak_idx = current_spectrum.findNearest(ann.mz, 1e-2);
+        if (peak_idx == -1) // no match
+        {
+          LOG_WARN << "Annotation present for missing peak. m/z: " << ann.mz
+                   << endl;
+          continue;
+        }
+        peak = current_spectrum[peak_idx];
+      }
+      ann_spectrum.push_back(peak);
+
+      String label = ann.annotation;
+      // write out positive and negative charges with the correct sign at the end of the annotation string
+      switch (ann.charge)
+      {
+      case 0: break;
+      case 1: label += "+"; break;
+      case 2: label += "++"; break;
+      case -1: label += "-"; break;
+      case -2: label += "--"; break;
+      default: label += ((ann.charge > 0) ? "+" : "") + String(ann.charge);
```

Comment:
can you change this so that it also works for negative mode spectra?

Reply:
I think that should work already. Or have I overlooked something? (In the default case, `String(ann.charge)` already contains the minus sign.)